py/makeqstrdata.py

```diff
@@ -344,19 +346,48 @@ def compute_huffman_coding(translations, compression_filename):
         # again, neither will "there" or "wither", since they have "the"
         # as substrings.
         extractor = TextSplitter(words)
+        counter = collections.Counter()
+        for t in texts:
+            for atom in extractor.iter(t):
+                counter[atom] += 1
+        cb = huffman.codebook(counter.items())
+        lengths = sorted(dict((v, len(cb[k])) for k, v in counter.items()).items())
+
+        def bit_length(s):
+            return sum(len(cb[c]) for c in s)
+
+        def est_len(occ):
+            idx = bisect.bisect_left(lengths, (occ, 0))
+            return lengths[idx][1] + 1
+
+        # The cost of adding a dictionary word is just its storage size
+        # while its savings is the difference between the original
+        # huffman bit-length of the string and the estimated bit-length
+        # of the dictionary word, times the number of times the word appears.
+        #
+        # The difference between the two is the net savings, in bits.
```

Comment:
Not strictly true because including a word into the Huffman tree bumps up the encoding lengths of all words in the same subtree. In the extreme case when the new word is so frequent that it gets a one-bit encoding, all other words will cost an extra bit each.

Reply:
Thanks, I adapted your explanation into an extended comment.